### ISSUE TITLE
getWordAtPosition not correct in JSON for long string values

### ISSUE THREAD

**Issue by aeschli** [microsoft/vscode#145019]
getWordAtPosition has an optimization to only look 15 characters before offset to figure out what the current word is.

This results in wrong words in case of JSON. The [JSON word rule](https://github.com/microsoft/vscode/blob/main/extensions/json/language-configuration.json#L18) is special as it (basically) says a word can start with `"` and end with `"`: `/"[^"]*"?/`

To reproduce:

- Have a JSON file with a long string value:
```json
    {
	"extensionTips": {
		"ms-dotnettools.csharp": "{**/*.cs,**/project.json,**/global.json,**/*.csproj,**/*.cshtml,**/*.sln,**/appsettings.json}"
        }
   }
```
- set a breakpoint in [getWordAtText](https://github.com/microsoft/vscode/blob/main/src/vs/editor/common/core/wordHelper.ts#L80)

- type a character (e/g/ comma) at the end of the line with the long string value

`getWordAtText` thinks the current word is `",`. If it had looked further ahead, it would have seen that it is not in a word.

**Comment by aeschli** [comment]
Yes, as explained, it matches to word that uses the closing `"` as the opening `"`

**Comment by jrieken** [comment]
It seems that the regex (`/("(?:[^\\\"]*(?:\\.)?)*"?)|[^\s{}\[\],:]+/g`) matches also on `",`. 

<img width="427" alt="Screenshot 2022-03-14 at 15 27 09" src="https://user-images.githubusercontent.com/1794099/158192876-611b0e41-1f75-4f01-97f5-3afe5087fc5d.png">

That satisfies `getWordAtText` logic because it has found a match that encloses the current position. I see how the result is bad but I am not really sure how to change the logic of `getWordAtText`. Do you suggest that it should keep finding a longer word - despite having matched already?

**Comment by aeschli** [comment]
Yes, I think it needs to be predictable how a line is broken up into words. Typically, scanner/tokenizers start at the beginning of the line and go incrementally from there.
I understand that this is problematic for very long lines, but this example isn't really one.

**Comment by jrieken** [comment]
> Yes, I think it needs to be predictable how a line is broken up into words. Typically, scanner/tokenizers start at the beginning of the line and go incrementally from there.

That's the point: it isn't broken up into words but starts at the wanted position. It's not a scanner/tokenizer and it doesn't start at the current line or document. Doing that would mean we start at offset zero and either advance by word-lengths or single character steps. That would be simply too slow...
